ocaml/lambda/debuginfo.ml

```diff
@@ -342,7 +342,12 @@ let to_location { dbg; assume_zero_alloc=_ } =
 
 let inline { dbg = dbg1; assume_zero_alloc = a1; }
       { dbg = dbg2; assume_zero_alloc = a2; } =
-  { dbg = dbg1 @ dbg2; assume_zero_alloc = ZA.Assume_info.meet a1 a2; }
+  { dbg = dbg1 @ dbg2;
+    assume_zero_alloc =
+      (* Drop "inferred" zero_alloc annotation from a call when
+         the callee is inlined. *)
+      if ZA.Assume_info.is_inferred a1 then a2 else
+      ZA.Assume_info.meet a1 a2; }
```

Comment:
I think this means the order of arguments to `inline` now matters, but didn't before.  Can we add a comment to the mli clarifying which argument is being inlined into the other?  Alternatively it might also be good to give one of the arguments a label making the order clear (e.g., `~into`?).  That way we'll be forced to look at all call sites and can verify they use the expected order.

I think the way this should work is that we look at the debug info of the function being inlined - if that one has an inferred za assume, we drop it.  I see a callsite in flambda like:
```ocaml
      Debuginfo.inline t.dbg dbg_from_inlined_body
```
Which makes me think we should be checking if the second one is inferred, not the first?  But I'm confused why this didn't break the tests if my understanding is correct.  It also looks like there is at least one caller in dead code under `ocaml/` for the old middle ends that uses the other order.

Reply:
I think the order mattered before, because of list concatenation in L345. I absolutely agree that the direction is confusing and I added a label (similarly to the one in flambd2).

Reply:
> I think the way this should work is that we look at the debug info of the function being inlined - if that one has an inferred za assume, we drop it. 

No, we look at the debug info on the call site: if it has "assume_inferred" annotation, then we use the annotation from the inlined body.

Reply:
Thanks, I agree I had this backwards yesterday.